File: frontend/js/maintenance.js
const API = "http://127.0.0.1:8000";

function headers() {
  return {
    "Content-Type": "application/json",
    Authorization: "Bearer " + localStorage.getItem("token"),
  };
}

async function addBook() {
  const mediaType = document.querySelector('input[name="media_type"]:checked')?.value || "book";
  const title = document.getElementById("title").value.trim();
  const author = document.getElementById("author").value.trim();
  const serialNo = document.getElementById("serial_no").value.trim();
  const category = document.getElementById("category").value.trim() || "general";
  const msg = document.getElementById("msg");

  if (!title || !author || !serialNo) {
    msg.innerText = "All fields are mandatory.";
    return;
  }

  const res = await fetch(`${API}/maintenance/add-book`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({
      media_type: mediaType,
      title,
      author,
      serial_no: serialNo,
      category,
    }),
  });
  const data = await res.json();
  msg.innerText = res.ok ? data.message : data.detail || "Add book failed";
}

async function addMembership() {
  const name = document.getElementById("name").value.trim();
  const duration = Number(document.querySelector('input[name="type"]:checked')?.value || "6");
  const msg = document.getElementById("msg");

  if (!name) {
    msg.innerText = "All fields are mandatory.";
    return;
  }

  const res = await fetch(`${API}/maintenance/add-membership`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({
      member_name: name,
      duration_months: duration,
    }),
  });
  const data = await res.json();
  msg.innerText = res.ok
    ? `Membership Created | Number: ${data.membership_number}`
    : data.detail || "Add membership failed";
}

async function updateMembership() {
  const number = document.getElementById("membership_number").value.trim();
  const action = document.querySelector('input[name="action"]:checked')?.value || "extend";
  const extension = Number(document.querySelector('input[name="extension"]:checked')?.value || "6");
  const msg = document.getElementById("msg");

  if (!number) {
    msg.innerText = "Membership number is mandatory.";
    return;
  }

  const res = await fetch(`${API}/maintenance/update-membership`, {
    method: "PUT",
    headers: headers(),
    body: JSON.stringify({
      membership_number: number,
      action,
      extension_months: extension,
    }),
  });
  const data = await res.json();
  msg.innerText = res.ok ? data.message : data.detail || "Update membership failed";
}

async function loadMemberships() {
  const res = await fetch(`${API}/maintenance/memberships`, { headers: headers() });
  const data = await res.json();
  const ul = document.getElementById("members");
  ul.innerHTML = "";
  if (!res.ok) return;
  data.forEach((m) => {
    ul.innerHTML += `<li class="list-group-item">${m.name} - ${m.type} - ${m.membership_number}</li>`;
  });
}

async function manageUser() {
  const mode = document.querySelector('input[name="mode"]:checked')?.value || "new";
  const name = document.getElementById("name").value.trim();
  const username = document.getElementById("username").value.trim();
  const password = document.getElementById("password").value.trim();
  const role = document.querySelector('input[name="role"]:checked')?.value || "user";
  const membershipNumber = document.getElementById("membership_number")?.value?.trim() || null;
  const msg = document.getElementById("msg");

  if (!name || !username) {
    msg.innerText = "Name and username are mandatory.";
    return;
  }
  if (mode === "new" && !password) {
    msg.innerText = "Password is mandatory for new user.";
    return;
  }

  const res = await fetch(`${API}/maintenance/user-management`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({
      mode,
      name,
      username,
      password: password || null,
      role,
      membership_number: membershipNumber,
    }),
  });
  const data = await res.json();
  msg.innerText = res.ok ? data.message : data.detail || "User management failed";
}

File: backend/app/routes/maintenance.py
from datetime import date, datetime, timedelta

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..auth import hash_password
from ..database import get_db
from ..dependencies import require_admin
from ..models import Book, Membership, User
from ..schemas import (
    BookCreateRequest,
    BookUpdateRequest,
    MembershipCreateRequest,
    MembershipUpdateRequest,
    UserManageRequest,
)

router = APIRouter(prefix="/maintenance", tags=["Maintenance"])


def _membership_number() -> str:
    return f"M-{date.today().strftime('%Y%m%d')}"


@router.post("/add-book")
def add_book(
    payload: BookCreateRequest,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    existing = db.query(Book).filter(Book.serial_no == payload.serial_no).first()
    if existing:
        raise HTTPException(status_code=400, detail="Duplicate serial number")

    book = Book(
        media_type=payload.media_type,
        title=payload.title.strip(),
        author=payload.author.strip(),
        serial_no=payload.serial_no.strip(),
        category=payload.category.strip(),
        available=True,
    )
    db.add(book)
    db.commit()
    db.refresh(book)
    return {"message": "Book added successfully", "book_id": book.id}


@router.put("/update-book")
def update_book(
    payload: BookUpdateRequest,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    book = db.query(Book).filter(Book.id == payload.book_id).first()
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")

    duplicate = (
        db.query(Book)
        .filter(Book.serial_no == payload.serial_no, Book.id != payload.book_id)
        .first()
    )
    if duplicate:
        raise HTTPException(status_code=400, detail="Duplicate serial number")

    book.media_type = payload.media_type
    book.title = payload.title.strip()
    book.author = payload.author.strip()
    book.serial_no = payload.serial_no.strip()
    book.category = payload.category.strip()
    book.available = payload.available
    db.commit()
    return {"message": "Book updated successfully"}


@router.post("/add-membership")
def add_membership(
    payload: MembershipCreateRequest,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    start_date = date.today()
    end_date = start_date + timedelta(days=payload.duration_months * 30)
    membership = Membership(
        membership_number=f"{_membership_number()}-{datetime.now().strftime('%H%M%S%f')}",
        name=payload.member_name.strip(),
        membership_type=f"{payload.duration_months}_months",
        start_date=start_date,
        end_date=end_date,
        active=True,
    )
    db.add(membership)
    db.commit()
    db.refresh(membership)
    return {
        "message": "Membership created successfully",
        "membership_number": membership.membership_number,
    }


@router.put("/update-membership")
def update_membership(
    payload: MembershipUpdateRequest,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    membership = (
        db.query(Membership)
        .filter(Membership.membership_number == payload.membership_number)
        .first()
    )
    if not membership:
        raise HTTPException(status_code=404, detail="Membership not found")

    if payload.action == "cancel":
        membership.active = False
        db.commit()
        return {"message": "Membership cancelled"}

    if not membership.active:
        raise HTTPException(status_code=400, detail="Cancelled membership cannot be extended")

    membership.end_date = membership.end_date + timedelta(days=payload.extension_months * 30)
    membership.membership_type = f"{payload.extension_months}_months"
    db.commit()
    return {"message": "Membership extended successfully"}


@router.get("/memberships")
def list_memberships(db: Session = Depends(get_db), _: User = Depends(require_admin)):
    memberships = db.query(Membership).all()
    return [
        {
            "name": m.name,
            "membership_number": m.membership_number,
            "type": m.membership_type,
            "start_date": m.start_date,
            "end_date": m.end_date,
            "active": m.active,
        }
        for m in memberships
    ]


@router.post("/user-management")
def user_management(
    payload: UserManageRequest,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    membership_id = None
    if payload.membership_number:
        membership = (
            db.query(Membership)
            .filter(Membership.membership_number == payload.membership_number)
            .first()
        )
        if not membership:
            raise HTTPException(status_code=404, detail="Membership not found")
        membership_id = membership.id

    existing = db.query(User).filter(User.username == payload.username).first()

    if payload.mode == "new":
        if existing:
            raise HTTPException(status_code=400, detail="Username already exists")
        if not payload.password:
            raise HTTPException(status_code=400, detail="Password is mandatory for new user")

        user = User(
            name=payload.name.strip(),
            username=payload.username.strip(),
            password=hash_password(payload.password),
            role=payload.role,
            membership_id=membership_id,
        )
        db.add(user)
        db.commit()
        db.refresh(user)
        return {"message": "User created successfully", "user_id": user.id}

    if not existing:
        raise HTTPException(status_code=404, detail="Existing user not found")

    existing.name = payload.name.strip()
    existing.role = payload.role
    existing.membership_id = membership_id
    if payload.password:
        existing.password = hash_password(payload.password)
    db.commit()
    return {"message": "User updated successfully", "user_id": existing.id}
